### DISCUSSION TITLE
Feature request: Add an execute method to SoapApi

### DISCUSSION THREAD

**Discussion by watzon** [jsforce/jsforce#1460] [Ideas]
The salesforce ajax toolkit has the ability to execute a function on a controller by calling the `execute` method with a controller, method, and parameter passed in. Something like that shouldn't be too hard to implement here following the same structure.

Some files for reference [connection.js and apex.js](https://gist.github.com/iDev0urer/ea77dca19638cbe975b3d25260f63025) which come directly from salesforce. The execute method from `apex.js` is the function that I want to replecate.

For reference purposes the xml structure that gets sent to salesforce looks like this:

``` xml
<se:Envelope xmlns:se="http://schemas.xmlsoap.org/soap/envelope/">
    <se:Header xmlns:sfns="http://soap.sforce.com/schemas/package/ControllerNamespace">
        <sfns:SessionHeader>
            <sessionId>00D36000000b28L!ARsAQLJW7bcHO</sessionId>
        </sfns:SessionHeader>
    </se:Header>
    <se:Body>
        <ListCategories xmlns="http://soap.sforce.com/schemas/package/ControllerNamespace" />
    </se:Body>
</se:Envelope>
```

I could really generate this myself if the SOAP class was exposed, but it seems to have been intentionally hidden.

_Update:_ After forking jsforce and exposing the SOAP class I was able to make this work with the following typescript:

``` typescript
private execute_rest(pkg: string, method: string, params: Object, callback?: Function): jsforce.Promise<any> {
        callback =  callback ? () => {} : callback;
        pkg = pkg.replace(/\./g, "/");

        let endpointUrl = this.conn.instanceUrl + '/services/Soap/package/' + pkg;
        let soap = new jsforce.SOAP(this.conn, {
            xmlns: `http://soap.sforce.com/schemas/package/${pkg}`,
            endpointUrl: endpointUrl
        });

      return soap.invoke(method, params, callback);
    }
```

This should be an extremely easy method to add.

**Comment by ivrtaric** [comment]
Hi @stomita,
I was wondering if this (or a similar solution) could be included in the 1.8.1 or 1.9.0 version of the jsforce npm package? Maybe not as a part of the Connection module - it looks as if it should belong in the SOAP API module.

@watzon 
Do you know if there's any difference between referencing the webservice container class using '/services/Soap/package/' and using '/services/Soap/class/'? From what I've seen, the first one is mainly referenced within Javascript Button handlers on the org itself, while the second one appears in the generated WSDL for the target class, but the behavior appears to be the same...

**Comment by watzon** [comment]
@ivrtaric no idea. To tell you the truth I haven't worked with Salesforce for the past several months so this project hasn't been in my tool bag

**Comment by apadak** [comment]
@watzon , how exactly did you expose the SOAP API? 
sorry. total noob here.
